JuliaStream.jl/src/AMDGPUStream.jl

```diff
@@ -0,0 +1,178 @@
+# AMDGPU.jl doesn't support CPU agents, so this isn't a feature-complete ROCmStream, only AMD GPUs
+include("Stream.jl")
+using AMDGPU
+
+const ROCData = StreamData{T,ROCArray{T}} where {T}
+const TBSize = 1024::Int
+const DotBlocks = 256::Int
+
+# AMDGPU.agents()'s internal iteration order isn't stable
+function gpu_agents_in_repr_order()
+  # XXX if we select anything other than :gpu, we get 
+  # HSA_STATUS_ERROR_INVALID_AGENT on the first kernel submission
+  sort(AMDGPU.get_agents(:gpu), by = repr)
+end
+
+function devices()
+  try
+    map(repr, gpu_agents_in_repr_order())
+  catch
+    # probably unsupported
+    []
+  end
+end
+
+function gridsize(data::ROCData{T})::Int where {T}
+  return data.size
+end
+
+function make_stream(
+  arraysize::Int,
+  scalar::T,
+  device::Int,
+  silent::Bool,
+)::ROCData{T} where {T}
+
+  if arraysize % TBSize != 0
+    error("arraysize ($(arraysize)) must be divisible by $(TBSize)!")
+  end
+
+  # XXX AMDGPU doesn't expose an API for setting the default like CUDA.device!()
+  # but AMDGPU.get_default_agent returns DEFAULT_AGENT so we can do it by hand
+  AMDGPU.DEFAULT_AGENT[] = gpu_agents_in_repr_order()[device]
+
+  data = ROCData{T}(
+    ROCArray{T}(undef, arraysize),
+    ROCArray{T}(undef, arraysize),
+    ROCArray{T}(undef, arraysize),
+    scalar,
+    arraysize,
+  )
+  selected = AMDGPU.get_default_agent()
+  if !silent
+    println("Using GPU HSA device: $(AMDGPU.get_name(selected)) ($(repr(selected)))")
+    println("Kernel parameters   : <<<$(gridsize(data)),$(TBSize)>>>")
+  end
+  return data
+end
+
+function hard_wait(kernel)
```

Comment:
Can we inline this so we don't introduce extra functions that look like they are APIs?

Reply:
Done